src/sinks/databend/service.rs

```diff
@@ -0,0 +1,217 @@
+    fn event_status(&self) -> EventStatus {
+        EventStatus::Delivered
+    }
+
+    fn events_sent(&self) -> CountByteSize {
+        CountByteSize(
+            self.metadata.event_count(),
+            self.metadata.events_estimated_json_encoded_byte_size(),
+        )
+    }
+
+    fn bytes_sent(&self) -> Option<usize> {
+        Some(self.metadata.request_encoded_size())
+    }
+}
+
+impl DatabendService {
+    pub(crate) const fn new(
+        client: DatabendAPIClient,
+        database: String,
+        table: String,
+    ) -> DatabendService {
+        DatabendService {
+            client,
+            database,
+            table,
+        }
+    }
+
+    pub(crate) fn new_stage_location(&self) -> Result<String, DatabendError> {
+        let now = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH)?;
+        let suffix = thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(8)
+            .map(char::from)
+            .collect::<String>();
+        Ok(format!(
+            "@~/vector/{}/{}/{}-{}",
+            self.database,
+            self.table,
+            now.as_secs(),
+            suffix,
+        ))
+    }
+
+    pub(crate) async fn get_presigned_url(
+        &self,
+        stage_location: &str,
+    ) -> Result<DatabendPresignedResponse, DatabendError> {
+        let req = DatabendHttpRequest::new(format!("PRESIGN UPLOAD {}", stage_location));
+        let resp = self.client.query(req).await?;
+
+        if resp.data.len() != 1 {
+            return Err(DatabendError::Server {
+                code: 500,
+                message: "Empty response from server for presigned request".to_string(),
+            });
+        }
+        if resp.data[0].len() != 3 {
+            return Err(DatabendError::Server {
+                code: 500,
+                message: "Invalid response from server for presigned request".to_string(),
+            });
+        }
+
+        // resp.data[0]: [ "PUT", "{\"host\":\"s3.us-east-2.amazonaws.com\"}", "https://s3.us-east-2.amazonaws.com/query-storage-xxxxx/tnxxxxx/stage/user/xxxx/xxx?" ]
+        let method = resp.data[0][0].clone();
+        let headers: BTreeMap<String, String> =
+            serde_json::from_str(resp.data[0][1].clone().as_str())?;
+        let url = resp.data[0][2].clone();
+
+        if method != "PUT" {
+            return Err(DatabendError::Server {
+                code: 500,
+                message: "Invalid method for presigned request".to_string(),
+            });
+        }
+
+        Ok(DatabendPresignedResponse {
+            method,
+            headers,
+            url,
+        })
+    }
+
+    pub(crate) async fn insert_with_stage(
+        &self,
+        stage_location: &str,
+    ) -> Result<(), DatabendError> {
+        let file_format_options = BTreeMap::from([("type".to_string(), "NDJSON".to_string())]);
+        let copy_options = BTreeMap::from([("purge".to_string(), "true".to_string())]);
```

Comment:
These options are always hardcoded, should they be static instead of cloning?

Reply:
made them as lazy load static